src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,355 @@
+                // User packets are only ever sent by an Awakener
+                return Ok(true);
+            }
+        };
+
+        let key = packet.key();
+        let (token, reg_type) = token_and_type_from_key(key);
+
+        match reg_type {
+            RegType::Handle => {
+                // We can return immediately-- no lookup or registration necessary.
+                evts.event_opt = Some(Event::new(signals_to_ready(observed_signals), token));
+                Ok(false)
+            },
+            RegType::Fd => {
+                // Convert the signals to epoll events using __mxio_wait_end,
+                // and add to reregistration list if necessary.
+                let events: u32;
+                {
+                    let handle = if let Some(handle) =
+                    self.token_to_fd
+                        .find(&Token(key as usize))
+                        .and_then(|h| h.get().upgrade()) {
+                        handle
+                    } else {
+                        // This handle is apparently in the process of removal.
+                        // It has been removed from the list, but port_cancel has not been called.
+                        return Ok(false);
+                    };
+
+                    events = unsafe {
+                        let mut events: u32 = mem::uninitialized();
+                        sys::fuchsia::sys::__mxio_wait_end(handle.mxio, observed_signals, &mut events);
+                        events
+                    };
+
+                    // If necessary, queue to be reregistered before next port_await
+                    let needs_to_rereg = {
+                        let registration_lock = handle.registration.lock().unwrap();
+
+                        registration_lock
+                            .as_ref()
+                            .map(|r| &r.rereg_signals)
+                            .is_some()
+                    };
+
+                    if needs_to_rereg {
+                        let mut tokens_to_rereg_lock = self.tokens_to_rereg.lock().unwrap();
+                        tokens_to_rereg_lock.push(token);
+                        self.has_tokens_to_rereg.store(true, Ordering::Relaxed);
+                    }
+                }
+
+                evts.event_opt = Some(Event::new(epoll_event_to_ready(events), token));
+                Ok(false)
+            },
+        }
+    }
+
+    /// Register event interests for the given IO handle with the OS
+    pub ( in sys::fuchsia) fn register_fd(&self,
+                                          handle: &magenta::Handle,
+                                          fd: &EventedFd,
+                                          token: Token,
+                                          signals: magenta::Signals,
+                                          poll_opts: PollOpt) -> io::Result<()>
+    {
+        self.token_to_fd.insert(token, Arc::downgrade(&fd.inner));
```

Comment:
I think it's technically valid on other platforms to register two handles with the same token, which seems like it could cause problems here as one would overwrite the other. In practice I don't think this happens in the ecosystem much though.